crates/phactory/src/lib.rs

```diff
@@ -542,6 +600,171 @@ impl<P: pal::Platform> Phactory<P> {
+        };
+        // Dispatch
+        let query_future = self
+            .system
+            .as_ref()
+            .expect("system always exists here")
+            .make_query(
+                req_id,
+                &AccountId::new(contract_id),
+                Some(&origin),
+                payload,
+                query_scheduler,
+                &self
+                    .runtime_state
+                    .as_ref()
+                    .expect("runtime state always exists here")
+                    .chain_storage,
+            );
+        let pink_runtime_version = self
+            .cluster_runtime_version()
+            .expect("BUG: no runtime version");
+        async move {
+            use phactory_api::contracts::QueryError;
+            use phala_types::contract::ContractQueryError;
+            use sidevm_env::messages::{QueryError as SidevmQueryError, QueryResponse};
+
+            fn opaqure_to_sidevm_err(err: ContractQueryError) -> SidevmQueryError {
+                match err {
+                    ContractQueryError::InvalidSignature => SidevmQueryError::InvalidSignature,
+                    ContractQueryError::ContractNotFound => SidevmQueryError::ContractNotFound,
+                    ContractQueryError::DecodeError => SidevmQueryError::DecodeError,
+                    ContractQueryError::OtherError(err) => SidevmQueryError::OtherError(err),
+                }
+            }
+            fn to_sidevm_err(err: QueryError) -> SidevmQueryError {
+                match err {
+                    QueryError::BadOrigin => SidevmQueryError::BadOrigin,
+                    QueryError::RuntimeError(err) => SidevmQueryError::RuntimeError(err),
+                    QueryError::SidevmNotFound => SidevmQueryError::SidevmNotFound,
+                    QueryError::NoResponse => SidevmQueryError::NoResponse,
+                    QueryError::ServiceUnavailable => SidevmQueryError::ServiceUnavailable,
+                    QueryError::Timeout => SidevmQueryError::Timeout,
+                }
+            }
+            let result: Result<QueryResponse, SidevmQueryError> = async move {
+                let (query_type, output, effects) = query_future
+                    .map_err(opaqure_to_sidevm_err)?
+                    .await
+                    .map_err(opaqure_to_sidevm_err)?;
+                let phactory_api::contracts::Response::Payload(output) =
+                    output.map_err(to_sidevm_err)?;
+                if let Some(effects) = effects {
+                    match weak_phactory.upgrade() {
+                        Some(phactory) => {
+                            phactory.lock().unwrap().apply_side_effects(effects);
+                        }
+                        None => {
+                            error!("Phactory dropped while processing sidevm query");
+                        }
+                    }
+                }
+                macro_rules! convert_result {
```

Comment:
For readability, it looks like we can move most of the code out from this function, and turn the macro into a normal function, since the only inputs are `pink_runtime_version` and `output`.

Reply:
To convert the macro into a function, we have to introduce a new trait to abstract over these two types:
```
    convert_result!(ContractExecResult)
    convert_result!(ContractInstantiateResult)
```
I feel this may not enhance readability compared to the declare macro but could add more complexity.
Perhaps my comfort with `macro_rules!` influences my perspective on this.

Reply:
Resolved: https://github.com/Phala-Network/phala-blockchain/pull/1398/commits/47c030be7b3a646ae211fd4f007a79dad2711448

Reply:
Nice. I think macro is ok, but embedding inside a function would make it super large, which reduces the readability.